// @vuetify/loader-shared/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.includes = exports.isObject = exports.resolveVuetifyBase = exports.writeStyles = exports.cacheDir = exports.generateImports = void 0;
const path = require("upath");
// | ((source: string, importer: string, isVuetify: boolean) => boolean | null | replace)
// type replace = { symbol: string, from: string, as?: string }
var generateImports_1 = require("./imports/generateImports");
Object.defineProperty(exports, "generateImports", { enumerable: true, get: function () { return generateImports_1.generateImports; } });
var writeStyles_1 = require("./styles/writeStyles");
Object.defineProperty(exports, "cacheDir", { enumerable: true, get: function () { return writeStyles_1.cacheDir; } });
Object.defineProperty(exports, "writeStyles", { enumerable: true, get: function () { return writeStyles_1.writeStyles; } });
function resolveVuetifyBase() {
    return path.dirname(require.resolve('vuetify/package.json', { paths: [process.cwd()] }));
}
exports.resolveVuetifyBase = resolveVuetifyBase;
function isObject(value) {
    return value !== null && typeof value === 'object';
}
exports.isObject = isObject;
function includes(arr, val) {
    return arr.includes(val);
}
exports.includes = includes;
